core/src/Kokkos_Parallel_Reduce.hpp

```diff
@@ -912,23 +912,50 @@ struct ReducerHasTestReferenceFunction {
   };
 };
 
-template <class ExecutionSpace, class T,
-          bool is_reducer = ReducerHasTestReferenceFunction<T>::value>
-struct ParallelReduceFence {
-  static void fence(const ExecutionSpace& execution_space, const T&) {
-    execution_space.fence();
-  }
-};
+template <class ExecutionSpace, class Arg>
+constexpr std::enable_if_t<
+    // constraints only necessary because SFINAE lacks subsumption
+    !ReducerHasTestReferenceFunction<Arg>::value &&
+        !Kokkos::is_view<Arg>::value,
+    // return type:
+    bool>
+parallel_reduce_needs_fence(ExecutionSpace const&, Arg const&) {
+  return true;
+}
+
+template <class ExecutionSpace, class Reducer>
+constexpr std::enable_if_t<
+    // equivalent to:
+    // (requires (Reducer const& r) {
+    //   { reducer.references_scalar() } -> std::convertible_to<bool>;
+    // })
+    ReducerHasTestReferenceFunction<Reducer>::value,
+    // return type:
+    bool>
+parallel_reduce_needs_fence(ExecutionSpace const&, Reducer const& reducer) {
+  return reducer.references_scalar();
+}
+
+template <class ExecutionSpace, class ViewLike>
+constexpr std::enable_if_t<
+    // requires Kokkos::ViewLike<ViewLike>
+    Kokkos::is_view<ViewLike>::value,
```

Comment:
I didn't want to use the token `View` because it shadows something in namespace Kokkos. But more to the point: in theory, things like `DynRankView` should also meet the requirements of `ViewLike` as a concept. Anyway, it's just a template parameter—don't read too much into it; I've just always thought of the concept for `View` as being named `ViewLike`, but do you have a better name for that concept?